libs/vkd3d-shader/spirv.c

```diff
@@ -1720,22 +1758,28 @@ static bool vkd3d_spirv_compile_module(struct vkd3d_spirv_builder *builder,
     vkd3d_spirv_build_word(&stream, 0); /* schema, reserved */
 
     /* capabilities */
-    for (i = 0; capability_mask; ++i)
-    {
-        if (capability_mask & 1)
-            vkd3d_spirv_build_op_capability(&stream, i);
-        capability_mask >>= 1;
-    }
-    if (builder->capability_draw_parameters)
-        vkd3d_spirv_build_op_capability(&stream, SpvCapabilityDrawParameters);
-    if (builder->capability_demote_to_helper_invocation)
-        vkd3d_spirv_build_op_capability(&stream, SpvCapabilityDemoteToHelperInvocationEXT);
+    for (i = 0; i < builder->capability_count; ++i)
+        vkd3d_spirv_build_op_capability(&stream, builder->capabilities[i]);
 
     /* extensions */
-    if (builder->capability_draw_parameters)
-        vkd3d_spirv_build_op_extension(&stream, "SPV_KHR_shader_draw_parameters");
-    if (builder->capability_demote_to_helper_invocation)
-        vkd3d_spirv_build_op_extension(&stream, "SPV_EXT_demote_to_helper_invocation");
+    for (i = 0; i < ARRAY_SIZE(vkd3d_spirv_capability_extensions); ++i)
+    {
+        if (vkd3d_spirv_has_capability(builder, vkd3d_spirv_capability_extensions[i].capability))
+        {
+            uint32_t extension = vkd3d_spirv_capability_extensions[i].extension;
+
+            if (extension_mask & extension)
```

Comment:
Oh, I see, it's using the bitmask to not enable the same extension over and over.